File: fullstack-roadmap/nivel-09-angular-backend/proyecto/frontend/src/app/services/producto.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Producto {
  id: number;
  nombre: string;
  descripcion: string;
  precio: number;
  existencia: number;
  categoriaId: number;
  sku: string;
  activo: boolean;
  categoriaNombre: string;
}

export interface ApiResponse<T> {
  success: boolean;
  data: T;
  total?: number;
  message?: string;
}

export interface PaginatedResponse {
  content: Producto[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
}

/**
 * Servicio HTTP: se comunica con la API de Spring Boot.
 * Usa HttpClient + Observables (RxJS).
 */
@Injectable({ providedIn: 'root' })
export class ProductoService {

  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/api/productos`;

  listarTodos(): Observable<ApiResponse<Producto[]>> {
    return this.http.get<ApiResponse<Producto[]>>(this.baseUrl);
  }

  buscarPorId(id: number): Observable<ApiResponse<Producto>> {
    return this.http.get<ApiResponse<Producto>>(`${this.baseUrl}/${id}`);
  }

  buscarPorNombre(nombre: string): Observable<ApiResponse<Producto[]>> {
    const params = new HttpParams().set('nombre', nombre);
    return this.http.get<ApiResponse<Producto[]>>(`${this.baseUrl}/buscar`, { params });
  }

  buscarPorCategoria(categoriaId: number): Observable<ApiResponse<Producto[]>> {
    return this.http.get<ApiResponse<Producto[]>>(`${this.baseUrl}/categoria/${categoriaId}`);
  }

  listarPaginado(page: number, size: number): Observable<PaginatedResponse> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PaginatedResponse>(`${this.baseUrl}/paginado`, { params });
  }

  crear(producto: Partial<Producto>): Observable<ApiResponse<Producto>> {
    return this.http.post<ApiResponse<Producto>>(this.baseUrl, producto);
  }

  actualizar(id: number, producto: Partial<Producto>): Observable<ApiResponse<Producto>> {
    return this.http.put<ApiResponse<Producto>>(`${this.baseUrl}/${id}`, producto);
  }

  eliminar(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.baseUrl}/${id}`);
  }
}

File: fullstack-roadmap/nivel-04-pruebas/proyecto/backend/src/main/java/com/softwarelee/productos/service/ProductoService.java
package com.softwarelee.productos.service;

import com.softwarelee.productos.mapper.ProductoMapper;
import com.softwarelee.productos.model.Producto;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

@Service
public class ProductoService {

    private final ProductoMapper productoMapper;

    public ProductoService(ProductoMapper productoMapper) {
        this.productoMapper = productoMapper;
    }

    public List<Producto> listarTodos() {
        return productoMapper.findAll();
    }

    public Producto buscarPorId(Long id) {
        Producto producto = productoMapper.findById(id);
        if (producto == null) {
            throw new RuntimeException("Producto no encontrado con ID: " + id);
        }
        return producto;
    }

    public List<Producto> buscarPorNombre(String nombre) {
        return productoMapper.findByNombre(nombre);
    }

    public List<Producto> buscarPorCategoria(Long categoriaId) {
        return productoMapper.findByCategoriaId(categoriaId);
    }

    public List<Producto> stockBajo(int minimo) {
        return productoMapper.findLowStock(minimo);
    }

    public Producto crear(Producto producto) {
        productoMapper.insert(producto);
        return productoMapper.findById(producto.getId());
    }

    public Producto actualizar(Long id, Producto producto) {
        buscarPorId(id);
        producto.setId(id);
        productoMapper.update(producto);
        return productoMapper.findById(id);
    }

    public void eliminar(Long id) {
        buscarPorId(id);
        productoMapper.softDelete(id);
    }

    public Map<String, Object> listarPaginado(int page, int size) {
        int offset = page * size;
        List<Producto> productos = productoMapper.findPaginated(offset, size);
        long total = productoMapper.count();
        int totalPages = (int) Math.ceil((double) total / size);

        return Map.of(
            "content", productos,
            "page", page,
            "size", size,
            "totalElements", total,
            "totalPages", totalPages
        );
    }
}
